core/src/main/scala/org/typelevel/keypool/internal/RequestSemaphore.scala

```diff
@@ -0,0 +1,121 @@
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package org.typelevel.keypool.internal
+
+import cats.syntax.all._
+import cats.effect.kernel.syntax.all._
+import cats.effect.kernel._
+import scala.collection.immutable.{Queue => ScalaQueue}
+import scala.annotation.nowarn
+
+import org.typelevel.keypool.Fairness
+
+/**
+ * RequestSemaphore moderates access to pooled connections by setting the number of permits
+ * available to the total number of connections. This is a custom semaphore implementation that only
+ * provides the `permit` operation. Additionally it takes a [[Fairness]] parameter, used to toggle
+ * the order in which requests acquire a permit.
+ *
+ * Derived from cats-effect MiniSemaphore
+ * https://github.com/typelevel/cats-effect/blob/v3.5.4/kernel/shared/src/main/scala/cats/effect/kernel/MiniSemaphore.scala#L29
+ */
+private[keypool] abstract class RequestSemaphore[F[_]] {
+  def permit: Resource[F, Unit]
+}
+
+private[keypool] object RequestSemaphore {
+  private trait BackingQueue[CC[_], A] {
+    def cleanup(fa: CC[A], elem: A): CC[A]
+    def offer(fa: CC[A], elem: A): CC[A]
+    def take(fa: CC[A]): (CC[A], A)
+    def nonEmpty(fa: CC[A]): Boolean
+  }
+
+  private implicit def listQueue[A <: AnyRef]: BackingQueue[List, A] = new BackingQueue[List, A] {
+    def cleanup(fa: List[A], elem: A) = fa.filterNot(_ eq elem)
+    def offer(fa: List[A], elem: A) = elem :: fa
+    def take(fa: List[A]) = (fa.tail, fa.head)
+    def nonEmpty(fa: List[A]) = fa.nonEmpty
+  }
+
+  private implicit def queueQueue[A <: AnyRef]: BackingQueue[ScalaQueue, A] =
+    new BackingQueue[ScalaQueue, A] {
+      def cleanup(fa: ScalaQueue[A], elem: A) = fa.filterNot(_ eq elem)
+      def offer(fa: ScalaQueue[A], elem: A) = fa :+ elem
+      def take(fa: ScalaQueue[A]) = (fa.tail, fa.head)
+      def nonEmpty(fa: ScalaQueue[A]) = fa.nonEmpty
+    }
+
+  private case class State[CC[_], A](waiting: CC[A], permits: Int)(implicit
+      @nowarn ev: BackingQueue[CC, A]
+  )
+
+  def apply[F[_]](fairness: Fairness, numPermits: Int)(implicit
+      F: GenConcurrent[F, _]
+  ): F[RequestSemaphore[F]] = {
+    require(numPermits >= 0, s"numPermits must be nonnegative, was: $numPermits")
```

Comment:
It's a nit, but this kind of `require` I'd prefer to see 
```suggestion
private def require(condition: Boolean, errorMessage: => String): F[Unit] =
if (condition) ().pure[F]
else new IllegalArgumentException(errorMessage).raiseError[F, Unit]

require(numPermits >= 0, s"numPermits must be nonnegative, was: $numPermits")
```

Reply:
I've done this!

Reply:
I *strongly* disagree. `IllegalArgumentException` (and by extension `Predef.require`) represents a programmer error and should never be handled<sup>†</sup>. in a correct program, `IllegalArgumentException` is never thrown. making it an effect treats it as a "normal" failure state which clients of the API should be aware of and potentially handle, which is incorrect

--------
<sup>†</sup> except when testing if it actually gets thrown

Reply:
I think I prefer the requested changes from `danicheg`. `IllegalArgumentException` isn't fatal, as in it's not one of the errors that will bypass the IO runtime, so I feel like using the error channel is reasonable. And it makes this apply referentially transparent which is I think a nice honest change. Also ... well..... I've definitely handled an `IllegalArgumentException` in my time, usually in the context of a web server and returning a 400 😅

The suggested approach feels more aligned with the rest of the codebase here

Reply:
This is a fun debate.  My kneejerk reaction is to raise, but @NthPortal makes a compelling argument.  Let's check cats-effect for prior art:

```scala
scala> Queue.bounded[IO, Int](-1)
java.lang.IllegalArgumentException: Bounded queue capacity must be non-negative, was: -1
  at cats.effect.std.Queue$.assertNonNegative(Queue.scala:176)
  at cats.effect.std.Queue$.bounded(Queue.scala:75)
  ... 32 elided
```

Okay, but:
       
```scala                                                                                                             
scala> val rand = cats.effect.std.Random.scalaUtilRandom[IO].unsafeRunSync()
val rand: cats.effect.std.Random[cats.effect.IO] = cats.effect.std.Random$$anon$2@6a622974
                                                                                                                                                                                                                                                    
scala> rand.nextIntBounded(-1)
val res2: cats.effect.IO[Int] = IO(...)
                                                                                                                          
scala> res2.unsafeRunSync()
java.lang.IllegalArgumentException: bound must be positive
  at java.base/java.util.Random.nextInt(Random.java:322)
  at nextIntBounded @ rs$line$9$.<clinit>(rs$line$9:1)
```

Fiddlesticks.

Is throwing an error pure?  Your friendly neighborhood Haskeller says yes:

```haskell
λ if (False) then error "oops" else ()
()
:: ()

λ let e = error "oops" in if (False) then e else ()
()
:: ()
```

Such errors are called [imprecise exceptions](https://www.tweag.io/blog/2020-04-16-exceptions-in-haskell/#imprecise-exceptions), and using them for programmer error resonates with @NthPortal's argument!  Rad.

Meanwhile, back in Scala,

```scala
scala> if (false) sys.error("oops") else ()
                                                                                                                          
scala> { val e = sys.error("oops"); if (false) e else () }
java.lang.RuntimeException: oops
  at scala.sys.package$.error(package.scala:27)
  ... 32 elided       
```

Extracting the errant expression to a value changes the result from a benign unit to a crash in Scala, due to strict evaluation.  But referential transparency depends on our point of observation.  Code that is impure _inside_ an `IO` is purified from a perspective _outside_ the `IO`.  Similarly, we can't observe the difference between the two expression above unless we:

- catch the exception.[^1]  
- crash and leave behind a stack trace and non-zero exit code.  But, lol, lmao, who cares, we're already dead!

The `require` approach shields the remainder of the program from this programmer error, at the expense of making the expression impure _to those who see the result_.  Haskellers and nihilists should prefer `require`, and on-call Scala programmers with yamlly configs could be better off with the `raiseError`.

[^1]: This is why Haskell's `catch` returns `IO`!